security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -253,52 +249,50 @@ func (sc *SecretManagerClient) GenerateSecret(resourceName string) (secret *secu
 		return ns, nil
 	}
 
-	if resourceName != RootCertReqResourceName {
-		// cache hit
-		if c := sc.cache.GetWorkload(); c != nil {
-			cacheLog.WithLabels("ttl", time.Until(c.ExpireTime)).Info("returned workload certificate from cache")
-			return c, nil
+	// cache hit. this is expected to happen most of the time.
+	if c := sc.cache.GetWorkload(); c != nil {
+		var rootCertBundle []byte
+		cacheLog.WithLabels("ttl", time.Until(c.ExpireTime)).Info("returned workload certificate from cache")
+		if resourceName == RootCertReqResourceName {
+			rootCertBundle = sc.mergeConfigTrustBundle(c.RootCert)
+		} else {
+			rootCertBundle = c.RootCert
 		}
-		t0 := time.Now()
-		ns, err := sc.generateSecret(resourceName)
-		if err != nil {
-			return nil, fmt.Errorf("failed to generate workload certificate: %v", err)
+		ns = &security.SecretItem{
+			ResourceName:     resourceName,
+			CertificateChain: c.CertificateChain,
+			PrivateKey:       c.PrivateKey,
+			RootCert:         rootCertBundle,
+			ExpireTime:       c.ExpireTime,
+			CreatedTime:      c.CreatedTime,
 		}
-
-		cacheLog.WithLabels("latency", time.Since(t0), "ttl", time.Until(ns.ExpireTime)).Info("generated new workload certificate")
-		sc.registerSecret(*ns)
 		return ns, nil
 	}
-	// If request is for root certificate,
-	// retry since rootCert may be empty until there is CSR response returned from CA.
```

Comment:
I don't think we should remove this. Instead, we should in addition make sure the root of trust we construct is always based on two sources.

Before we send the root SDS response, I think we need to make sure 1) the root from the CSR response is received 2) the root list from the XDS sidechannel is received (may be empty). Otherwise, the response will be partial and incorrect.

Reply:
The code above handles what happens if the workload certificate response is present in the cache (both are merged). The code below handles what happens if there is a cache miss - we trigger and wait for the CSR response and merge with the proxyConfig certs. 
Most importantly, when the ProxyConfigCerts are updated for any reason, we trigger a callback that forces  a ROOTCA request from envoy